Editorial review: Document web app manifest localization (#44609)
mdn/content: files/en-us/web/progressive_web_apps/how_to/localize_an_app_manifest/index.md

---
title: Localize an app manifest
slug: Web/Progressive_web_apps/How_to/Localize_an_app_manifest
page-type: how-to
sidebar: pwasidebar
---

> 

Reviewer: How do I localise the rest of the app? We should at least mention that (or if it can't be done, state that too). I assume perhaps we can specify a key for the default language entry point?

Author: Localizing the rest of the app is a huge topic and out of scope for this article. How you would go about localizing it depends on what kind of app it is, what stack you've used, etc.

https://github.com/oh-jon-paul/awesome-i18n gives you an idea of what's available.

> [Progressive Web App (PWA)](/en-US/docs/Web/Progressive_web_apps) manifests can be localized using a combination of the [`*_localized`](/en-US/docs/Web/Progressive_web_apps/Manifest/Reference/*_localized) manifest member suffix, the [`lang`](/en-US/docs/Web/Progressive_web_apps/Manifest/Reference/lang) member, and the [`dir`](/en-US/docs/Web/Progressive_web_apps/Manifest/Reference/dir) member.

Reviewer: its not the `*_localized`  manifest member suffix, but "members with the suffix". We should drop the `*` in that name. Perhaps

```suggestion
[Progressive Web App (PWA)](/en-US/docs/Web/Progressive_web_apps) manifests can be localized using a combination of localizable members, which have the suffix [`_localized`](/en-US/docs/Web/Progressive_web_apps/Manifest/Reference/*_localized), and the [`lang`](/en-US/docs/Web/Progressive_web_apps/Manifest/Reference/lang) and [`dir`](/en-US/docs/Web/Progressive_web_apps/Manifest/Reference/dir) members.
```

Manifest localization allows the browser to choose between different text strings (for app names and descriptions, for example) and icons when loading the app to best suit the user's language preferences.

> This guide shows how to localize a PWA manifest, using the following example as a starting point:

Reviewer: It feels a bit abrupt to jump into this template, and also that it doesn't belong in the intro. Just a suggestion, but you might break this like this, and provide a description of this default.

```suggestion
This guide shows how to localize a PWA manifest.

## Example non-localised PWA

The rest of this guide develops a multilingual PWA manifest using the following example as a starting point.
```

And we might also include non-localizable elements (if not already) so we can explain them too.

Author: I've removed the last sentence from your suggestion because it is not entirely true. I've not included `shortcuts`. I didn't think it was necessary to show all possible localizable features, as it would make the article longer and more repetitive.

I don't think it is necessary to spend time explaining the non-localizable members. That seems out of scope to me.

```json
{
  "name": "The SuperSausage sausage app",
  "short_name": "SuperSausage",
  "description": "Find information on all your favorite sausages!",
  "start_url": "./",
  "display": "standalone",
  "background_color": "#ffffff",
  "theme_color": "#ab510d",
  "icons": [
    {
      "src": "./icons/saus-128.png",
      "sizes": "128x128",
      "type": "image/png"
    },
    {
      "src": "./icons/saus-256.png",
      "sizes": "256x256",
      "type": "image/png"
    }
  ]
}
```

## Set a default language and direction

> The [`lang`](/en-US/docs/Web/Progressive_web_apps/Manifest/Reference/lang) and [`dir`](/en-US/docs/Web/Progressive_web_apps/Manifest/Reference/dir) members define a default language and language direction for the app. These will be assumed by the browser if no language and direction more suitable for the user's preferences are found in the [`*_localized`](/en-US/docs/Web/Progressive_web_apps/Manifest/Reference/*_localized) variants. In such cases, the non-prefixed members are used (for example, `name`).

Reviewer: 1. What if I don't set a default language or dir, what is used? 
2. While I understand exactly what you mean, this doesn't feel quite right

   > _These_ will be assumed by the browser if no language and direction more suitable for the user's preferences are found in the [`*_localized`](/en-US/docs/Web/Progressive_web_apps/Manifest/Reference/*_localized) variants.

   It's because the languages will be used but it isn't `lang` and `dir` that will be used. What will be used is the other non-localized members. You clarify this in the next sentence, but "These" still grates.

Reviewer 2: 1. At least in the Chromium implementation, if no `dir` is specified for a localized member, it uses the direction of the browser's currently set language. Localizations are selected based off of the browser's currently set language at all times.

Author: See previous comments.

Reviewer: @kyerebo Thanks. What's the point then? I mean you only get the localised strings if your browser is set to a localisation and if you haven't set the `dir` you'll be inheriting the correct direction from your browser. 

Or to put it another way, say I've set arabic in my browser. I will get RTL and arabic strings by default, so there's no need to set `dir` unless I want LTR - why would I want to do that when reading arabic?

The answer to this affects things like "why we specified dir for urdu in this example": https://github.com/mdn/content/pull/44609/changes#r3575092791

Reviewer 2: After double-checking the implementation, while Chromium pumps the localized text object's `dir` and `lang` fields to the web app's internal storage, there's actually no user-visible place where they are being consumed. 
Like you said, with the language being set in the browser, the correct direction is handled by the browser itself. 

In practice, this means that there's not really much point to having a localized text object `dir` field (at least for Chromium, uncertain how other UAs may implement this.) For `lang`, it could maybe serve as a notation for anyone that's looking at the manifest to say, "this value is in a different language than the tag indicates". 

In terms of the example, perhaps it'd be clearer to remove usages of localized object `dir` tags.

Author: Ah, perfect, thanks for the info. This being the case, I have:

- Completely removed any mention of the `dir` member from the howto guide.
- Reduced mention of it on the reference page, and updated the necessary mentions to make it clear that the language direction is inferred from the browser language setting and that you'll most commonly want to just use the string form.

- [`icons`](/en-US/docs/Web/Progressive_web_apps/Manifest/Reference/icons)
- [`shortcuts`](/en-US/docs/Web/Progressive_web_apps/Manifest/Reference/shortcuts)

Our sample manifest has all of these except for `shortcuts`; we'll show how to localize each one.

## Choose the languages you want to support

Generally you should support all of the languages relevant to your major target audience locales.

For this example, we'll choose French (`fr`), German (`de`), Urdu (`ur`), and Japanese (`ja`).

## Localizing the `name` member

Localized versions of the `name` member are contained in the `name_localized` member. Each property inside `name_localized` has to have a key equal to one of the target language's {{glossary("BCP 47 language tag", "BCP 47 language tags")}}, and a value equal to an object or a string representing the localized text for that language.

Let's see what this could look like for our example:

```json
"name_localized": {
  "fr": "L'application de saucisse SuperSausage",
  "de": "Die SuperWurst-App",
  "ur": {
    "dir": "rtl",
    "value": "سپر ساسیج ساسیج ایپ"
  },
  "ja": "スーパーソーセージのソーセージアプリ"
}
```

> In this case, we've used the string form for all the languages, except for Urdu, which uses the object form. This is because Urdu is an RTL language — we have specified a `dir` value of `rtl` for it, so that browsers will display it correctly.

Reviewer: ```suggestion
In this case, we've used the string form for all the languages, except for Urdu, which uses the object form. We've done it this way because Urdu is an RTL language, and we need to specify a `dir` value of `rtl` for it so that browsers will display it correctly.
```

## Localizing the other text-based members

We can follow the same pattern as we did for the `name` member to localize the `short_name` and `description` members:

```json
"short_name_localized": {
  "fr": {
    "lang": "en-US",
    "value": "SuperSausage"
  },
  "de": "SuperWurst",
  "ur": "سپر ساسیج",
  "ja": "スーパーソーセージ"
},
"description_localized": {
  "fr": "Trouvez des informations sur toutes vos saucisses préférées !",
  "de": "Finden Sie Informationen zu all Ihren Lieblingswürstchen!",
  "ur": "اپنی تمام پسندیدہ ساسیجز کے بارے میں معلومات حاصل کریں!",
  "ja": "お気に入りのソーセージの情報を全部見つけましょう!"
},
```

> The French (`fr`) `short_name` translation shows typical usage of the object value form being used to specify a `lang` property. In this case, our French audience knows our app by its English brand name — "SuperSausage" — and we want to specify that this should be handled as English rather than French (for example, for the purposes of pronounciation).

Reviewer: See https://github.com/mdn/content/pull/44609/changes#r3517481036 - `lang` in a translation needs additional explanation there.  I think I get it now, finally on this the second last doc - this allows you to specify that the value from some other language key should be used for a translation, instead of providing your own, or using the default?

Typo too

```suggestion
The French (`fr`) `short_name` translation shows typical usage of the object value form being used to specify a `lang` property. In this case, our French audience knows our app by its English brand name — "SuperSausage" — and we want to specify that this should be handled as English rather than French (for example, for the purposes of pronunciation).
```

Reviewer 2: I'm reading it as clarification that although the localized object is under the French language tag, the value of the object is presented in English rather than French.

From the web application manifest spec:
> To support multilingual content and ensure optimal display and accessibility, it is possible to specify a different language for a [localized text object](https://www.w3.org/TR/appmanifest/#dfn-localized-text-object). This is needed for situations where a term or text needs to be presented in a language different from the user's set locale.

Author: see above

Author: I've fixed the typo

## Localizing the `icon` member

> The `icon` localization is covered in a separate section because localized images are handled somewhat differently to localized text. The properties of each `*_localized` member (`icon_localized`, in this case) are equal to an array of objects containing the same properties as the non-localized original (`icon`); each one provides details for a localized image.

Reviewer: They don't seem so different. Maybe

```suggestion
Localized images are handled somewhat differently to localized text. The properties of each `*_localized` member (`icon_localized`, in this case) are equal to an array of objects containing the same properties as the non-localized original (`icon`); each one provides details for a localized image.
```

Let's see what this looks like:

```json
"icons_localized": {
  "fr": [
    {
      "src": "./icons/l10n/fr/saus-128.png",
      "sizes": "128x128",
      "type": "image/png"
    },
    {
      "src": "./icons/l10n/fr/saus-256.png",
      "sizes": "256x256",
      "type": "image/png"
    }
  ],
  "de": [
    {
      "src": "./icons/l10n/de/saus-128.png",
      "sizes": "128x128",
      "type": "image/png"
    },
    {
      "src": "./icons/l10n/de/saus-256.png",
      "sizes": "256x256",
      "type": "image/png"
    }
  ],
  "ur": [
    {
      "src": "./icons/l10n/ur/saus-128.png",
      "sizes": "128x128",
      "type": "image/png"
    },
    {
      "src": "./icons/l10n/ur/saus-256.png",
      "sizes": "256x256",
      "type": "image/png"
    }
  ],
  "ja": [
    {
      "src": "./icons/l10n/ja/saus-128.png",
      "sizes": "128x128",
      "type": "image/png"
    },
    {
      "src": "./icons/l10n/ja/saus-256.png",
      "sizes": "256x256",
      "type": "image/png"
    }
  ],
}
```

## Finished manifest

> Putting this all together, the complete manifest looks like this:

Reviewer: I think you need your shortcuts - I want to know how to select my language version.

Author: Do you mean you want me to include a `shortcuts` example?

---
title: Localize an app manifest
slug: Web/Progressive_web_apps/How_to/Localize_an_app_manifest
page-type: how-to
sidebar: pwasidebar
---

> [Progressive Web App (PWA)](/en-US/docs/Web/Progressive_web_apps) manifests can be localized using a combination of localizable members, which have the suffix [`_localized`](/en-US/docs/Web/Progressive_web_apps/Manifest/Reference/*_localized), and the [`lang`](/en-US/docs/Web/Progressive_web_apps/Manifest/Reference/lang) and [`dir`](/en-US/docs/Web/Progressive_web_apps/Manifest/Reference/dir) members.

Reviewer 2: Touched upon this in email with Chris, reposting here in more detail.
The `dir` and `lang` manifest-level members are somewhat independent of the localization feature.
`*_localized` members are selected based off of the browser's language setting, not off of the other manifest members.
This is definitely a bit confusing in the spec, as it says the manifest-level `dir/lang` set the default value of localized members, but the section on localized members clarifies that the default representation is the manifest member without `*_localized` and localizations should be selected based off of the [user's language setting](https://www.w3.org/TR/appmanifest/#dfn-default-representation).
The manifest-level `dir` is defined in the spec as applying a default text direction for localized members that are missing a `dir` field, but the Chromium implementation only prioritizes whatever direction is specified by the localized member.
The manifest-level `lang` field isn't actually used when calculating which localized value to show. 

In practice (or at least how it works right now in Chromium), this means that the manifest-level `lang` and `dir` don't have implications for the localized value that ends up getting chosen. Perhaps we can remove or edit the references to these manifest level members to avoid confusing developers?
The suggested instructions for working with localizable manifests on the chrome developer blog don't mention manifest-level `dir` or `lang` as any sort of fallback.

Author: I think I've captured this all in my updates, except maybe this bit:

> The manifest-level dir is defined in the spec as applying a default text direction for localized members that are missing a dir field, but the Chromium implementation only prioritizes whatever direction is specified by the localized member.

But I'm not sure if there is really much worth saying.

Reviewer 2: I agree it's probably not worth specifying. LGTM.

Reviewer: See my comments. IMO we should completely purge the `lang` from the docs until there is an implementation.

My understanding for `dir` from what you say above is that
-  in spec the default for `dir` should be set in the top level, and then might be overridden in a localisation
- But in chrome the manifest/top level dir is ignored (and presumably `dir` is guessed from the language itself) but can be overridden by the localisation.

I'd say that perhaps we should remove this too, though the use-case is more arguable since it does do something.

Author: To reiterate — the `lang` and `dir` top-level members are the ones that are not implemented anywhere. `lang` and `dir` inside the textual `*_localized` members are implemented.

With the localized members, e.g. `name_localized`, one of the provided options will be chosen if the language tag key matches the user's browser language setting. If none of the provided options match the user's browser language setting, the non-localized member will be used, e.g. `name`.

## Example non-localised PWA

The rest of this guide develops a multilingual PWA manifest using the following example as a starting point.

```json
{
  "name": "The SuperSausage sausage app",
  "short_name": "SuperSausage",
  "description": "Find information on all your favorite sausages!",
  "start_url": "./",
  "display": "standalone",
  "background_color": "#ffffff",
  "theme_color": "#ab510d",
  "icons": [
    {
      "src": "./icons/saus-128.png",
      "sizes": "128x128",
      "type": "image/png"
    },
    {
      "src": "./icons/saus-256.png",
      "sizes": "256x256",
      "type": "image/png"
    }
  ]
}
```

## Set a default language and direction

The [`lang`](/en-US/docs/Web/Progressive_web_apps/Manifest/Reference/lang) and [`dir`](/en-US/docs/Web/Progressive_web_apps/Manifest/Reference/dir) members define a default language and language direction for the app. These will be assumed by the browser if no language and direction more suitable for the user's preferences are found in the [`*_localized`](/en-US/docs/Web/Progressive_web_apps/Manifest/Reference/*_localized) variants. In such cases, the non-prefixed members are used (for example, `name`).

> Set a suitable default language and direction like so:

Reviewer 2: Consider removing/editing this section as part of comment above.

Author: done

- [`name`](/en-US/docs/Web/Progressive_web_apps/Manifest/Reference/name)
- [`short_name`](/en-US/docs/Web/Progressive_web_apps/Manifest/Reference/short_name)
- [`description`](/en-US/docs/Web/Progressive_web_apps/Manifest/Reference/description)
- [`icons`](/en-US/docs/Web/Progressive_web_apps/Manifest/Reference/icons)
- [`shortcuts`](/en-US/docs/Web/Progressive_web_apps/Manifest/Reference/shortcuts)

Our sample manifest has all of these; we'll show how to localize each one.

## Choose the languages you want to support

Generally you should support all of the languages relevant to your major target audience locales.

For this example, we'll choose French (`fr`), German (`de`), Urdu (`ur`), and Japanese (`ja`).

## Localizing the `name` member

Localized versions of the `name` member are contained in the `name_localized` member. Each property inside `name_localized` has to have a key equal to one of the target language's {{glossary("BCP 47 language tag", "BCP 47 language tags")}}, and a value equal to an object or a string representing the localized text for that language.

Let's see what this could look like for our example:

```json
"name_localized": {

>   "fr": "L'application de saucisse SuperSausage",

Reviewer: I still think it would be useful to have one of these as a more complicated language-region tag as you did previously. The reason being that users might wish to support regions, and they might further like to understand the fallback approach. Non blocking though.

Author: This raises a good question — if the user has a general language set for their browser (e.g. `fr`) and the app has a regional variation available (e.g. `fr-CA`) — or vice versa — does the browser use that variation even though it is not identical, language tag-wise? The spec wording is "best matches the user's preference".

Do you have an answer for this, @kyerebo?

I think this is kind of what you are getting at, @hamishwillee. If we can get an answer to this question, we can add a regional example and provide guidance on this at the same time.

Reviewer 2: The Chromium implementation matches the more granular language tag first, then falls back. For example, if the user's browser is set to `fr-CA`, it'll look for localizations with the language tag of `fr-CA`, and if that fails, look for `fr`, and if that fails, fall back on the non-localized value.

Reviewer: Cool. Yes, this is what I meant (and what I expected).

Author: Brilliant, thanks a lot for the info, @kyerebo. I've added a note to the reference page to cover this, and updated the howto page to include a `fr-CA` variant for the `icons`, plus explanations of what `fr-CA` users will see used in each case.

  "de": "Die SuperWurst-App",
  "ur": {

>     "dir": "rtl",

Reviewer: Didn't you remove this?

Author: See previous comment.

    "value": "سپر ساسیج ساسیج ایپ"
  },
  "ja": "スーパーソーセージのソーセージアプリ"
}
```

> In this case, we've used the string form for all the languages, except for Urdu, which uses the object form. We've done it this way because Urdu is an RTL language, and we need to specify a `dir` value of `rtl` for it so that browsers will display it correctly.

Reviewer: Dir here too. We should note that the directionality comes from the language itself.

Reviewer: @chrisdavidmills OK, so I have just gone almost all the way through this. I based my review on 

> OK, shortcuts localization details added, and dir and lang member pages removed (as they are not actually supported anywhere yet).

To me this says we shouldn't document them at all. We should explain how it works now. If we are going to document them now we need to document it fully, but we'd still need to explain how it works now.

I'll have another look after you respond to this one.

Author: See previous comment.

> In this case, we've used the string form for all the languages, except for Urdu, which uses the object form. We've done it this way because Urdu is an RTL language, and we need to specify a `dir` value of `rtl` for it so that browsers will display it correctly.

Reviewer: I don't believe you needed to set `dir` here - if your browser was set to Urdu you would get the correct directionality anyway. If it wasn't, then you didn't get to display this string. Right?

This depends on answer to https://github.com/mdn/content/pull/44609#discussion_r3575000116 - maybe there is a reason why you "need" to set directionality.

Author: Yeah, this is a good point. Let's see what @kyerebo says.

Reviewer 2: @hamishwillee is correct here, no need to set directionality as per my recent response in https://github.com/mdn/content/pull/44609#discussion_r3581281981

If the user has their browser language set to `fr`, `de`, `ur`, or `ja`, the browser will use the appropriate name found in the `name_localized` member for that language as the app's `name`. If not, the browser will use the name found in the `name` member.

## Localizing the other text-based members

We can follow the same pattern as we did for the `name` member to localize the `short_name` and `description` members:

```json
"short_name_localized": {
  "fr": {

>     "lang": "en-US",

Reviewer: You removed lang. Also mentioned below and should be purged.

Author: See previous

    "value": "SuperSausage"
  },
  "de": "SuperWurst",
  "ur": "سپر ساسیج",
  "ja": "スーパーソーセージ"
},
"description_localized": {
  "fr": "Trouvez des informations sur toutes vos saucisses préférées !",
  "de": "Finden Sie Informationen zu all Ihren Lieblingswürstchen!",
  "ur": "اپنی تمام پسندیدہ ساسیجز کے بارے میں معلومات حاصل کریں!",
  "ja": "お気に入りのソーセージの情報を全部見つけましょう!"
},
```

> The French (`fr`) `short_name` translation shows typical usage of the object value form with a `lang` property specified. In this case, our French audience knows our app by its English brand name — "SuperSausage" — and we want to specify that this should be handled as English rather than French (for example, for the purposes of pronunciation).

Reviewer: We might still have a comment though. I.e. what would you actually do now since you can't specify the lang?

Author: See previous

> The French (`fr`) `short_name` translation shows typical usage of the object value form with a `lang` property specified. In this case, our French audience knows our app by its English brand name — "SuperSausage" — and we want to specify that this should be handled as English rather than French (for example, for the purposes of pronunciation).

Reviewer: What would the difference have been if we just omitted `fr` from `short_name_localized`? As I understand it we'd just fall back to the English name, so why would we do it this way?

Or to put it another way, I understand what this does, but I am not sure of "why we bother".

Reviewer: Actually, I do see a use. It might not be English we're using for our default. Perhaps we want to fall back to a Spanish name (say) rather than all the way back to English, or we want to go to a more specific version of French and we don't want to redeclare.

Probably worth capturing the nuance. Maybe with a different example than `en-US` which is not declared anywhere.

Author: I get what you mean, but I feel like the examples you've hinted at seem pretty niche. It would also be a huge pain to include a whole new example just to explain this.

I've had a go at handling this by specifying the "French English" as a variant of the English brand name, rather than the same English brand name (which would be pointless). Let me know what you think.

## Localizing the `icons` member

> Localized icons are handled somewhat differently to localized text. The properties of each `icons_localized` member are equal to an array of objects containing the same properties as the non-localized `icons` member; each one provides details for a localized icon.

Reviewer: I'd say they are handled the same except that each value is an array of objects rather than a string. What I do want to know here is if the arrays must match the `icons` arrays in terms of sizes and number of elements, or are the arrays completely independent. For example, an I declare 20 icons for English and just one for french, and if I do, what will the browser select - just those in the fr key? Or the best size match?

Author: Good question. Any thoughts, @kyerebo?

Reviewer 2: Chromium browsers will treat each localized icons array as completely independent. 
In that example, as far as a French localized app is concerned, there would only be one icon available to choose from in terms of sizes.

Author: Cool, good to know, thanks @kyerebo!

I've added a paragraph to cover this inside the reference page:

> Each localized icons array is treated as completely independent from all the others. If an icons variant matches the user's browser language setting, only icons from within that variant will be chosen for that user. For example, if you have 20 icons specified inside `icons`, and only one icon specified inside `icons_localized.fr`, users with `fr` set as their browser's language will only ever see one icon used everywhere. The browser won't look inside the `icons` array for more suitable sizes.

    }
  ],
  "de": [
    {
      "src": "./icons/l10n/de/saus-128.png",
      "sizes": "128x128",
      "type": "image/png"
    },
    {
      "src": "./icons/l10n/de/saus-256.png",
      "sizes": "256x256",
      "type": "image/png"
    }
  ],
  "ur": [
    {
      "src": "./icons/l10n/ur/saus-128.png",
      "sizes": "128x128",
      "type": "image/png"
    },
    {
      "src": "./icons/l10n/ur/saus-256.png",
      "sizes": "256x256",
      "type": "image/png"
    }
  ],
  "ja": [
    {
      "src": "./icons/l10n/ja/saus-128.png",
      "sizes": "128x128",
      "type": "image/png"
    },
    {
      "src": "./icons/l10n/ja/saus-256.png",
      "sizes": "256x256",
      "type": "image/png"
    }
  ],
}
```

## Localizing the `shortcuts` member

In the case of the `shortcuts` member, you don't specify the localizations inside a `shortcuts_localized` member. Instead, you provide `*_localized` versions of the `name`, `short_name`, `description`, and `icons` members nested inside the `shortcut` member.

For our example, this looks like so:

```json
"shortcuts": [
  {
    "name": "Open menu",
    "name_localized": {
      "fr": "Menu ouvert",
      "de": "Menü öffnen",

>       "ur": { "value": "اوپن مینو", "dir": "rtl" },

Reviewer: Same as https://github.com/mdn/content/pull/44609/changes#r3575092791 - though I guess no need to explain twice.

      "ja": "メニューを開く"
    },
    "short_name": "Menu",
    "short_name_localized": {
      "fr": "Menu",
      "de": "Speisekarte",

>       "ur": { "value": "مینو", "dir": "rtl" },

Reviewer: What about `dir` here?

Author: See previous
